### ISSUE TITLE
naming-convention Distinguish between public object and class properties

### ISSUE THREAD

**Issue by twastvedt** [typescript-eslint/typescript-eslint#1895]
**Repro**

```JSON
{
  "rules": {
    "@typescript-eslint/naming-convention": [
      "error",
      {
        "selector": "property",
        "modifiers": ["public"],
        "format": ["PascalCase"]
      }
    ]
  }
}
```

```TS
class Test {
  public Id = 8;
  private o = {
    id: 8,
  };
}
```

**Expected Result**
Works based on documentation, but I would like an option to distinguish between public class properties and public object properties so that the example does not cause errors.

**Actual Result**
Error on the object field `Test.o.id` requiring PascalCase.

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `2.28.0` |
| `@typescript-eslint/parser`        | `2.28.0` |
| `TypeScript`                       | `3.8.3` |
| `ESLint`                           | `6.8.0` |
| `node`                             | `10.17.0` |
| `npm`                              | `6.11.3` |

Thanks!

**Comment by bradzacher** [comment]
Out of curiosity - why do you name class properties in PascalCase, and object properties in camelCase? It seems like an odd distinction considering that TS treats variables with a class type the same way as an object (because classes are essentially syntactic sugar on top of objects)

This will very much a non-trivial change to make, and the only way to do it may be via a breaking change. Happy if you want to have a go at it!

**Comment by twastvedt** [comment]
Interesting point, and I'm definitely open to suggestion. Still trying to figure out our naming convention. Coming from C# I've found the conventional distinction between PascalCase public properties and camelCase private properties to be useful. I suppose one could argue that then all object properties should be PascalCase, since they're all, technically, public. But as far as I've seen, that would be pretty contrary to Javascript convention, which I'm also trying to keep in mind. 

Do Typescript objects have the concept of private properties? As far as I know, they don't. If that's the case, then even if behind the scenes they're the same as classes, in front of the scenes they aren't. Naming conventions seem like an in-front-of-the-scenes thing.

Regardless of the resolution of the above discussion, which I'm definitely interested in, it's good to know this wouldn't be trivial. Unfortunately I don't have the bandwidth for coding up a non-trivial PR.

**Comment by bradzacher** [comment]
Outside of typescript, the naming convention that generally exists is camelCase all the things, private members have a single underscore prefix, and protected members have a double underscore prefix.

Some important notes:
- Unlike in C#, PascalCase is only really used for class names, and type names.
- The above underscore naming convention exists for one reason: because there is no such thing as accessibility checks in pure JS.  
In a world without accessibility enforcement, you have to trust engineers to not use private members. Having the simple underscore naming convention makes it easy to grep for during code reviews: "oh there's a `._`, that shouldn't be in this code <Request Changes>".
- A number of codebases use this naming convention to help build chains identify private members so they can be transformed to be truly private.
Eg.
```js
class Foo {
  publicMember = 1;
  _privateMember = 2;
  __protectedMember = 3;
}
// transforms to
var Foo = function Foo() {
  "use strict";
  this.publicMember = 1;
  this.$Foo_privateMember = 2;
  this.__protectedMember = 3;
};
```

Moving on to typescript code. In TS code, you don't need to worry about catching this sort of thing at review time, or via a build transform because the TS compiler will fail on code ignores the accessibility. There are only two ways around this build error: a `// @ts-ignore` comment, which stands out in a code review (and can be banned further with our lint rule `ban-ts-comment`), or via string index access: `x['privateMember']`, which again stands out like a sore thumb - and can be blocked via the core lint rule `dot-notation`.

So personally, I think there is near-zero value in applying a naming convention to specific modifiers in typescript code.

If your class is part of a public API intended to be consumed from pure JS code, then I'd go with the underscore-based naming convention, as it's what people generally expect.  
OTOH, if your class is part of a public API intended to be consumed only from TS code (i.e. it's a package internal to your codebase), then I wouldn't bother with a naming convention, as typescript will generate a definition file _with_ the accessibility modifiers in it, preventing people from misusing the class.

----

phew.
With all that said, there's actually another option, which was added in TS3.8 - private class members (a part of the official JS spec).

```ts
class Foo {
  #trulyPrivateMember = 1;
}

const x = new Foo();
x.#trulyPrivateMember;
// Property '#trulyPrivateMember' is not accessible outside class 'Foo' because it has a private identifier.
```
[repl](https://www.typescriptlang.org/v2/en/play?target=2#code/MYGwhgzhAEBiD29oG8BQ1oGIAuAnAriAJ4AKuAlgG5jYCmAsrQLYBGtu0AvNAIwDcqAL5A)

when your target is less than `esnext`, private class properties will be transpiled down to keep their true private-ness, and it's not just a compile-time error to attempt to access one - it _will_ be a runtime error.

Why is there a difference between the `private` modifier and `#` private members? This is intentional for a few reasons, namely changing the meaning of `private` would be a _huge_ breaking change across the ecosystem.

**Comment by twastvedt** [comment]
Thanks for the detailed intro! I think we'll move away from needing this for our conventions.
